include/nbl/video/alloc/SubAllocatedDescriptorSet.h

```diff
@@ -248,76 +220,60 @@ class SubAllocatedDescriptorSet : public core::IReferenceCounted
-
-			unallocatedSize = try_multi_allocate(
-				binding, 
-				unallocatedSize, 
-				&descriptors[count - unallocatedSize], 
-				&outDescriptorWrites[count - unallocatedSize], 
-				&outAddresses[count - unallocatedSize]
-			);
+			eventHandler.wait(maxWaitPoint, unallocatedSize, outNulls);
+
+			// always call with the same parameters, otherwise this turns into a mess with the non invalid_address gaps
+			unallocatedSize = try_multi_allocate(binding,count,outAddresses);
 			if (!unallocatedSize)
-				return 0u;
+				break;
 		} while(Clock::now()<maxWaitPoint);
+		m_logicalDevice->nullifyDescriptors({nulls.data(),outNulls});
 
 		return unallocatedSize;
 	}
 
-	inline size_type multi_allocate(
-		uint32_t binding, 
-		size_type count, 
-		video::IGPUDescriptorSet::SDescriptorInfo* descriptors, 
-		video::IGPUDescriptorSet::SWriteDescriptorSet* outDescriptorWrites, 
-		value_type* outAddresses
-	) noexcept
+	// default timeout overload
+	inline size_type multi_allocate(const uint32_t binding, const size_type count, value_type* outAddresses) noexcept
 	{
-		auto range = m_allocatableRanges.find(binding);
-		// Check if this binding has an allocator
-		if (range == m_allocatableRanges.end())
-			return count;
-
-		return multi_allocate(TimelineEventHandlerBase::default_wait(), binding, count, descriptors, outDescriptorWrites, outAddresses);
+		// check that the binding is allocatable is done inside anyway
+		return multi_allocate(TimelineEventHandlerBase::default_wait(), binding, count, outAddresses);
 	}
 
-	inline void multi_deallocate(uint32_t binding, size_type count, const size_type* addr)
+	// Very explicit low level call you'd need to sync and drop descriptors by yourself
+	// Returns: the one-past the last `outNullify` write pointer, this allows you to work out how many descriptors were freed
+	inline void multi_deallocate(IGPUDescriptorSet::SDropDescriptorSet* outNullify, uint32_t binding, size_type count, const size_type* addr)
 	{
 		auto debugGuard = stAccessVerifyDebugGuard();
 
 		auto allocator = getBindingAllocator(binding);
-		if (!allocator)
-			return;
+		if (allocator)
 		for (size_type i=0; i<count; i++)
 		{
 			if (addr[i]==AddressAllocator::invalid_address)
 				continue;
 
 			allocator->free_addr(addr[i],1);
-			// TODO: should also write something to the descriptor sets
-			// basically if nullDescriptor device feature is enabled, you would 
-			// indeed write to the DS, else you'd just drop the refcounted references
-			//
-			// this needs to be done as a IGPUDescriptorSet::nullify(const uint32_t binding, 
-			// std::span<uint32_t> indices) function + a virtual nullify_impl
-			video::IGPUDescriptorSet::SDropDescriptorSet dropDescriptorSet;
-			dropDescriptorSet.dstSet = m_descriptorSet.get();
-			dropDescriptorSet.binding = binding;
-			dropDescriptorSet.arrayElement = i;
-			dropDescriptorSet.count = 1;
-			m_logicalDevice->nullifyDescriptors({ &dropDescriptorSet, 1 });
```

Comment:
calling it for a single descriptor at a time would have tanked your perf if you actually had the nullDescriptor feature enabled